### PR TITLE
requires_grad=False when no_grad()

### PR DESCRIPTION
**Pull request by ssnl**
This allows native functions to do computation conditioning on whether gradient is needed. 

For example, SVD backward needs U and V tensors, but if we just need the singular values and the decomposed tensor doesn't require grad, we can safely skip computing U and V. Empirically, this speeds up nuclear norm calculation for tenors with `requires_grad=False` a lot.

```py
# CUDA_LAUNCH_BLOCKING=1

x = (torch.randn(1000, 1000) + torch.eye(1000).div_(20)).cuda()

%timeit -r 100 x.norm('nuc')

# pre patch: 21.3 ms ± 1.99 ms per loop (mean ± std. dev. of 100 runs, 1 loop each)
# post patch: 9.77 ms ± 295 µs per loop (mean ± std. dev. of 100 runs, 1 loop each)

x.requires_grad_()

%timeit -r 100 x.norm('nuc')

# pre patch: 25.4 ms ± 2.24 ms per loop (mean ± std. dev. of 100 runs, 1 loop each)
# post patch: 24.9 ms ± 1.46 ms per loop (mean ± std. dev. of 100 runs, 1 loop each)

x = (torch.randn(1000, 1000) + torch.eye(1000).div_(20)).cpu()

%timeit -r 100 x.norm('nuc')

# pre patch: 3.73 ms ± 396 µs per loop (mean ± std. dev. of 100 runs, 1 loop each)
# post patch: 1.86 ms ± 52.7 µs per loop (mean ± std. dev. of 100 runs, 1 loop each)

x.requires_grad_()

%timeit -r 100 x.norm('nuc')

# pre patch: 3.3 ms ± 478 µs per loop (mean ± std. dev. of 100 runs, 1 loop each)
# post patch: 3.55 ms ± 434 µs per loop (mean ± std. dev. of 100 runs, 1 loop each)
```

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -558,7 +558,9 @@ Tensor nuclear_norm(const Tensor& self, bool keepdim) {
       "Expected a tensor with 2 dimensions, but got a ",
       self.dim(),
       " dimensions tensor instead.");
-  return at::sum(std::get<1>(at::svd(self)), 0, keepdim);
+  bool self_need_grad = self.is_variable() && self.requires_grad();
+  auto S = std::get<1>(at::svd(self, /*some=*/true, /*compute_uv=*/self_need_grad));
```
Comment (vishwakftw) [inline aten/src/ATen/native/LinearAlgebra.cpp]: Wow, this is amazing!!

File: torch/csrc/autograd/functions/utils.h [hunk 2]
```diff
@@ -42,9 +42,6 @@ struct ComputeRequiresGrad : IterArgs<ComputeRequiresGrad> {
 
 template <typename... Args>
 inline bool compute_requires_grad(Args&&... args) {
-  if (!GradMode::is_enabled()) {
```
Comment (albanD) [inline torch/csrc/autograd/functions/utils.h]: Why is this removed?
It looks like this function is used in most auto-generated backward, is there a performance impact on adding more indirection to it?

Comment (SsnL) [reply torch/csrc/autograd/functions/utils.h]: The idea is to move `GradMode` check to `Varaible::requires_grad`, so we can use it in ATen. Good point on the performance, but given that our functions usually take 1~4 inputs, I think extra 2~3 bool checks per function is not a big problem :)

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -1671,7 +1673,8 @@ Tensor slogdet_backward(const std::vector<torch::autograd::Variable> &grads,
     return grads[1] * self.inverse().t();
   } else /* otherwise det = \prod(sigma) = 0, use svd */ {
     Tensor u, sigma, v;
-    std::tie(u, sigma, v) = self.svd();
+    bool self_need_grad = self.is_variable() && self.requires_grad();
+    std::tie(u, sigma, v) = self.svd(/*some=*/true, /*compute_uv=*/self_need_grad);
```
Comment (apaszke) [inline tools/autograd/templates/Functions.cpp]: This is really confusing for me. You’re computing those values only conditionally, but you’re always using them below??

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 4]
```diff
@@ -567,7 +569,9 @@ Tensor &nuclear_norm_out(Tensor& result, const Tensor& self, bool keepdim) {
       "Expected a tensor with 2 dimensions, but got a ",
       self.dim(),
       " dimensions tensor instead.");
-  return at::sum_out(result, std::get<1>(at::svd(self)), 0, keepdim);
+  bool self_need_grad = self.is_variable() && self.requires_grad();
+  auto S = std::get<1>(at::svd(self, /*some=*/true, /*compute_uv=*/self_need_grad));
```
Comment (apaszke) [inline aten/src/ATen/native/LinearAlgebra.cpp]: I really don’t get it. Why do you have to specify the value of this argument in the forward function? Is running this function with compute_uv false an error if the input requires grad or what?

File: tools/autograd/templates/Functions.cpp (hunk 3 above)
Comment (SsnL) [reply tools/autograd/templates/Functions.cpp]: omg. solid point. These should always compute u and v. I'll revert these changes. Thanks!

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 4 above)
Comment (SsnL) [reply aten/src/ATen/native/LinearAlgebra.cpp]: `compute_uv` is required to be `true` if one wants to backprop through `svd` (yes it's an error otherwise). So if the input of forward don't require grad, we can skip it because we only use `s`.

File: torch/csrc/autograd/functions/utils.h (hunk 2 above)
Comment (colesbury) [reply torch/csrc/autograd/functions/utils.h]: Huh? So variable.requires_grad() will return false if GradMode isn't enabled even if the variable requires grad? That doesn't seem like a good idea

Comment (colesbury) [reply torch/csrc/autograd/functions/utils.h]: The short-circuit works the other way. If GradMode is off, then it'll have to check *every* input as currently written in this PR.

But the more important issue is that this changes the semantics of requires_grad(), which I don't think we should be doing.

Comment (SsnL) [reply torch/csrc/autograd/functions/utils.h]: @colesbury Yes, that is indeed the part I am not so sure about, but it would enable optimizations that would be otherwise impossible because `GradMode` lives in `torch::`.

Comment (colesbury) [reply torch/csrc/autograd/functions/utils.h]: Can't you move GradMode out of torch without changing the semantics of requires_grad?
